src/main/scala/com/dwolla/http4s/consul/ResourceMap.scala

```diff
@@ -0,0 +1,67 @@
+package com.dwolla.http4s.consul
+
+import cats.data.OptionT
+import cats.effect._
+import cats.effect.syntax.all._
+import cats.syntax.all._
+import io.chrisdavenport.mapref.MapRef
+import org.typelevel.log4cats.Logger
+
+/**
+ * A data structure to map keys to (1) an effect that can view a most recent value
+ * for the key and (2) a background process that refreshes the most recent value
+ * over time.
+ *
+ * Because the data structure tracks background processes, it's expected that
+ * instances of [[ResourceMap]] will themselves only be available inside a
+ * `cats.effect.Resource` managing the lifecycle of that [[ResourceMap]] and
+ * its background processes.
+ */
+sealed trait ResourceMap[F[_], K, V] {
+  def get(k: K): F[V]
+}
+
+object ResourceMap {
+  def apply[F[_] : Async : Logger, K, V](startRefreshingK: K => Resource[F, GetCurrentValue[F, V]]): Resource[F, ResourceMap[F, K, V]] =
+    makeMapRefAndKeyList[F, K, (GetCurrentValue[F, V], F[Unit])]
+      .toResource
+      .flatMap { case (mapRef, keys) =>
+        Resource.pure {
```

Comment:
This is the block I'm most unsure of. 

1. Is it sound to do `mapRef(k).get.flatMap(v => … mapRef(k).set)`?
2. Are there cancelation or other gotchas when calling `resource.allocated` and then storing the finalizer in the `MapRef` for use when the `Resource[F, ResourceMap[F, K, V]]` is finalized?

Reply:
The core idea here is that, over time, code using this data structure will (indirectly, via calls to `ResourceMap.get`) call the `K => Resource[F, GetCurrentValue[F, V]]` function with different `K` values. Each distinct `K` will have its own `Resource[F, GetCurrentValue[F, V]]`, and we want to:

1. Tie the finalization of those resources to the lifecycle of the outer `Resource[F, ResourceMap[F, K, V]]`
2. Only call the `K => Resource[F, GetCurrentValue[F, V]]` once per unique `K`. (Subsequent calls should just use the `F[V]` inside the `GetCurrentValue[F, V]` newtype.)